statesync/blocks_test.go

```diff
@@ -0,0 +1,248 @@
+package statesync
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tendermint/internal/test/factory"
+	"github.com/tendermint/tendermint/p2p"
+)
+
+var (
+	startHeight int64 = 200
+	stopHeight  int64 = 100
+	stopTime          = time.Date(2019, 1, 1, 1, 0, 0, 0, time.UTC)
+	endTime           = stopTime.Add(-1 * time.Second)
+	numWorkers        = 1
+)
+
+func TestBlockQueueBasic(t *testing.T) {
+	peerID, err := p2p.NewNodeID("0011223344556677889900112233445566778899")
+	require.NoError(t, err)
+
+	queue := newBlockQueue(startHeight, stopHeight, stopTime, 1)
+	wg := &sync.WaitGroup{}
+
+	// asynchronously fetch blocks and add it to the queue
+	for i := 0; i <= numWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			for {
+				select {
+				case height := <-queue.NextHeight():
+					queue.Add(mockLBResp(t, peerID, height, endTime))
+				case <-queue.Done():
+					wg.Done()
+					return
+				}
+			}
+		}()
+	}
+
+	trackingHeight := startHeight
+	wg.Add(1)
+
+loop:
+	for {
+		select {
+		case <-queue.Done():
+			wg.Done()
+			break loop
+
+		case resp := <-queue.VerifyNext():
+			// assert that the queue serializes the blocks
+			assert.Equal(t, resp.block.Height, trackingHeight)
+			trackingHeight--
+			queue.Success(resp.block.Height)
+		}
+
+	}
+
+	wg.Wait()
+	assert.Less(t, trackingHeight, stopHeight)
+
+	select {
+	case <-queue.doneCh:
+	default:
+		t.Fatal("queue's done channel is not closed")
```

Comment:
I think this can never happen because in order to get past line 66, line 54 has to run.